### PR TITLE
[SPARK-33865][SPARK-36202][SQL]  When HiveDDL, we need check avro schema too

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Unify schema check code of FileFormat and check avro schema filed name when CREATE TABLE DDL too

### Why are the changes needed?
Refactor code

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Not need

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
gentle ping @dongjoon-hyun @HyukjinKwon @cloud-fan @xkrogen

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 1]
```diff
@@ -933,19 +934,34 @@ object DDLUtils {
...
-        case "parquet" => ParquetSchemaConverter.checkFieldNames(schema)
-        case "orc" => OrcFileFormat.checkFieldNames(schema)
+        case "parquet" => checkDataColNames("parquet", schema)
+        case "orc" => checkDataColNames("orc", schema)
         case _ =>
       }
     }
   }
 
+  private[sql] def checkDataColNames(provider: String, schema: StructType): Unit = {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:950]:
```suggestion
  def checkDataColNames(provider: String, schema: StructType): Unit = {
```

execution package is already private. let's remove

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala [hunk 2]
```diff
@@ -112,4 +112,21 @@ trait FileDataSourceV2 extends TableProvider with DataSourceRegister {
...
+   */
+  def checkFieldName(name: String): Unit = {}
+
+  def checkFieldNames(schema: StructType): Unit = {
+    schema.foreach { field =>
+      checkFieldName(field.name)
+      field.dataType match {
+        case s: StructType => checkFieldNames(s)
+        case _ =>
+      }
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala:129]: I don't have a preference. Just one thought:

```suggestion
      if (field.dataType.isInstanceOf[StructType]) checkFieldNames(field.dataType.asInstanceOf[StructType])
```

up to your preference

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala [hunk 3]
```diff
@@ -112,4 +112,21 @@ trait FileDataSourceV2 extends TableProvider with DataSourceRegister {
...
   }
+
+  /**
+   * Check whether target schema field name is valid.
+   *
+   * @throws AnalysisException If field name is invalid
+   */
+  def checkFieldName(name: String): Unit = {}
+
+  def checkFieldNames(schema: StructType): Unit = {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala:123]: Actually, I wouldn't add this as a method in fact ..

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 4]
```diff
@@ -933,19 +934,34 @@ object DDLUtils {
...
       }
     }
   }
 
+  def checkDataColNames(provider: String, schema: StructType): Unit = {
+    try {
+      DataSource.lookupDataSource(provider, SQLConf.get).getConstructor().newInstance() match {
+        case f: FileFormat => DataSourceUtils.verifySchema(f, schema)
+        case f: FileDataSourceV2 =>
+          f.getTable(schema, Array.empty, Map.empty[String, String].asJava).schema()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:955]: nit: `DataSourceUtils.verifySchema(f.fallbackFileFormat, schema)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 5]
```diff
@@ -933,19 +934,34 @@ object DDLUtils {
...
+        case _ =>
+      }
+    } catch {
+      case e: AnalysisException if e.getMessage.contains("contains invalid character") => throw e
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:959]: it's a bit hacky to check on the error message here. how about
```
val source = try {
  DataSource.lookupDataSource(provider, SQLConf.get).getConstructor().newInstance()
} catch {
  case NonFatal(e) =>
    logError...
    return
}

source match ...

```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala [hunk 6]
```diff
@@ -68,6 +78,7 @@ object DataSourceUtils {
         throw QueryCompilationErrors.dataTypeUnsupportedByDataSourceError(format.toString, field)
       }
     }
+    checkFieldNames(format, schema)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala:81]: is this change needed?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala (hunk 4 above)
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala (hunk 5 above)
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Done

**Comment by AngersZhuuuu** [comment]
retest this please

**Comment by AngersZhuuuu** [comment]
Any more suggestion?

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala [hunk 7]
```diff
@@ -153,6 +154,27 @@ private[sql] class AvroFileFormat extends FileFormat
   }
 
   override def supportDataType(dataType: DataType): Boolean = AvroUtils.supportsDataType(dataType)
+
+  override def supportFieldName(name: String): Unit = {
```
Comment (cloud-fan) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala:158]: can it return boolean?

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala [hunk 8]
```diff
@@ -153,6 +154,27 @@ private[sql] class AvroFileFormat extends FileFormat
...
+    val length = name.length
+    if (length == 0) {
+      throw QueryCompilationErrors.columnNameContainsInvalidCharactersError(name)
+    } else {
+      val first = name.charAt(0)
+      if (!Character.isLetter(first) && first != '_') {
```
Comment (cloud-fan) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala:164]: do you have some reference doc to prove this is indeed a limitation of avro? or can you run some local tests like `df.write.format("avro").save(path)`?

Comment (AngersZhuuuu) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: > do you have some reference doc to prove this is indeed a limitation of avro? or can you run some local tests like `df.write.format("avro").save(path)`?

For `df.write.format(source).save(path)`

Avro error message
```
[info]   org.apache.avro.SchemaParseException: Illegal initial character: (IF((ID = 1), 1, 0))
[info]   at org.apache.avro.Schema.validateName(Schema.java:1562)
[info]   at org.apache.avro.Schema.access$400(Schema.java:91)
[info]   at org.apache.avro.Schema$Field.<init>(Schema.java:546)
[info]   at org.apache.avro.SchemaBuilder$FieldBuilder.completeField(SchemaBuilder.java:2240)
[info]   at org.apache.avro.SchemaBuilder$FieldBuilder.completeField(SchemaBuilder.java:2236)
[info]   at org.apache.avro.SchemaBuilder$FieldBuilder.access$5100(SchemaBuilder.java:2150)
[info]   at org.apache.avro.SchemaBuilder$GenericDefault.noDefault(SchemaBuilder.java:2539)
[info]   at org.apache.spark.sql.avro.SchemaConverters$.$anonfun$toAvroType$1(SchemaConverters.scala:194)
[info]   at scala.collection.Iterator.foreach(Iterator.scala:943)
[info]   at scala.collection.Iterator.foreach$(Iterator.scala:943)
[info]   at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
[info]   at scala.collection.IterableLike.foreach(IterableLike.scala:74)
[info]   at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
[info]   at org.apache.spark.sql.types.StructType.foreach(StructType.scala:102)
[info]   at org.apache.spark.sql.avro.SchemaConverters$.toAvroType(SchemaConverters.scala:191)
[info]   at org.apache.spark.sql.avro.AvroUtils$.$anonfun$prepareWrite$1(AvroUtils.scala:98)
[info]   at scala.Option.getOrElse(Option.scala:189)
[info]   at org.apache.spark.sql.avro.AvroUtils$.prepareWrite(AvroUtils.scala:97)
[info]   at org.apache.spark.sql.avro.AvroFileFormat.prepareWrite(AvroFileFormat.scala:74)
[info]   at org.apache.spark.sql.execution.datasources.FileFormatWriter$.write(FileFormatWriter.scala:142)
[info]   at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelationCommand.run(InsertIntoHadoopFsRelationCommand.scala:186)
[info]   at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult$lzycompute(commands.scala:113)
[info]   at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult(commands.scala:111)
[info]   at org.apache.spark.sql.execution.command.DataWritingCommandExec.executeCollect(commands.scala:125)
[info]   at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.$anonfun$applyOrElse$1(QueryExecution.scala:97)
[info]   at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
[info]   at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
[info]   at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
[info]   at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775)
```

parquet error message
```
[info] - SPARK-33865: Hive DDL with avro should check col name *** FAILED *** (4 seconds, 554 milliseconds)
[info]   org.apache.spark.sql.AnalysisException: Column name "(IF((ID = 1), 1, 0))" contains invalid character(s). Please use alias to rename it.
[info]   at org.apache.spark.sql.errors.QueryCompilationErrors$.columnNameContainsInvalidCharactersError(QueryCompilationErrors.scala:2105)
[info]   at org.apache.spark.sql.execution.datasources.parquet.ParquetSchemaConverter$.checkFieldName(ParquetSchemaConverter.scala:590)
[info]   at org.apache.spark.sql.execution.datasources.parquet.ParquetWriteSupport$.$anonfun$setSchema$2(ParquetWriteSupport.scala:485)
[info]   at org.apache.spark.sql.execution.datasources.parquet.ParquetWriteSupport$.$anonfun$setSchema$2$adapted(ParquetWriteSupport.scala:485)
[info]   at scala.collection.immutable.List.foreach(List.scala:431)
[info]   at org.apache.spark.sql.execution.datasources.parquet.ParquetWriteSupport$.setSchema(ParquetWriteSupport.scala:485)
[info]   at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat.prepareWrite(ParquetFileFormat.scala:110)
[info]   at org.apache.spark.sql.execution.datasources.FileFormatWriter$.write(FileFormatWriter.scala:142)
[info]   at org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelationCommand.run(InsertIntoHadoopFsRelationCommand.scala:186)
[info]   at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult$lzycompute(commands.scala:113)
[info]   at org.apache.spark.sql.execution.command.DataWritingCommandExec.sideEffectResult(commands.scala:111)
[info]   at org.apache.spark.sql.execution.command.DataWritingCommandExec.executeCollect(commands.scala:125)
[info]   at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.$anonfun$applyOrElse$1(QueryExecution.scala:97)
[info]   at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
[info]   at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
[info]   at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
[info]   at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775)
[info]   at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
[info]   at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:97)
[info]   at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:93)
[info]   at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:481)
[info]   at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:82)
[info]   at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:481)
```

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala (hunk 7 above)
Comment (AngersZhuuuu) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: How about current?

**Comment by AngersZhuuuu** [comment]
retest this please

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala [hunk 9]
```diff
@@ -51,9 +51,9 @@ private[sql] class AvroFileFormat extends FileFormat
   override def hashCode(): Int = super.hashCode()
 
   override def inferSchema(
-      spark: SparkSession,
-      options: Map[String, String],
-      files: Seq[FileStatus]): Option[StructType] = {
+    spark: SparkSession,
```
Comment (cloud-fan) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala:54]: the new indentation is wrong

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala [hunk 10]
```diff
@@ -58,6 +58,22 @@ object DataSourceUtils {
     Serialization.read[Seq[String]](str)
   }
 
+  /**
+   * Verify if the field name is supported in datasource. This verification should be done
+   * in a drover side.
+   */
+  def checkFieldNames(format: FileFormat, schema: StructType): Unit = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala:65]: shall we merge this into `verifySchema`? The `schema: StructType` contains field names as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala [hunk 11]
```diff
@@ -163,6 +163,14 @@ trait FileFormat {
    * By default all data types are supported.
    */
   def supportDataType(dataType: DataType): Boolean = true
+
+
+  /**
+   * Check whether target schema field name is valid.
+   *
+   * @throws AnalysisException If field name is invalid
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:171]: we should remove this now

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala [hunk 12]
```diff
@@ -58,6 +58,22 @@ object DataSourceUtils {
     Serialization.read[Seq[String]](str)
   }
 
+  /**
+   * Verify if the field name is supported in datasource. This verification should be done
+   * in a drover side.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala:63]: in the driver side

Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala]: DOne

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala (hunk 11 above)
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala]: Done

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala (hunk 9 above)
Comment (AngersZhuuuu) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: Done

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala [hunk 13]
```diff
@@ -2158,6 +2158,36 @@ abstract class AvroSuite
       }
     }
   }
+
+  test("SPARK-33865: Hive DDL with avro should check col name") {
```
Comment (cloud-fan) [inline external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:2162]: why do we mention Hive? The test uses `USING AVRO` which is Spark native CREATE TABLE DDL.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala [hunk 14]
```diff
@@ -163,6 +163,12 @@ trait FileFormat {
    * By default all data types are supported.
    */
   def supportDataType(dataType: DataType): Boolean = true
+
+
+  /**
+   * Check whether target schema field name is valid.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:169]: can we follow the doc of the method above?

**Review by cloud-fan — APPROVED** [review]
LGTM. Please also mention the avro change in the PR description

File: external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala (hunk 13 above)
Comment (AngersZhuuuu) [reply external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:2162]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala (hunk 14 above)
Comment (AngersZhuuuu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala:169]: Done

**Comment by cloud-fan** [comment]
The failure in GA is unrelated and jenkins passes.

thanks, merging to master/3.2!
